### ISSUE TITLE
Capturing output from WASI binary (produced by TinyGo 0.26.0)

### ISSUE THREAD

**Issue by thisisaaronland** [bytecodealliance/wasmtime-py#115]
I'd like to use a WASI binary produced by TinyGo (0.26.0) in Python capturing the output from an exported function to a variable. 

Specifically the `parse` method defined here: 

https://github.com/sfomuseum/go-edtf/blob/wasi/cmd/parse-wasi/main.go

Which I can build and test like this:

```
$> tinygo build -o parse.wasm -target wasi ./cmd/parse-wasi/main.go

$> wasmtime parse.wasm 2022-01-01
2022-01-01
```

_At this point the Go/WASM binary just parrots back input pending some decisions about how to work around parts of the Go stdlib that aren't available in TinyGo yet._

If I run the following Python code (in a file called `test.py`):

```
#!/usr/local/bin/python3

from wasmtime import Store, Module, Instance, Func, FuncType

if __name__ == "__main__":

    store = Store()
    module = Module.from_file(store.engine, "parse.wasm")

    instance = Instance(store, module, [])
    parse = instance.exports(store)["parse"]

    print(parse("2022-01-01"))
```

It fails with the following error:

```
Traceback (most recent call last):
  File "/usr/local/sfomuseum/go-edtf/./test.py", line 10, in <module>
    instance = Instance(store, module, [])
  File "/usr/local/lib/python3.10/site-packages/wasmtime/_instance.py", line 43, in __init__
    raise WasmtimeError._from_ptr(error)
wasmtime._error.WasmtimeError: expected 14 imports, found 0
``` 

I feel like I am missing something obvious but I am not sure what, exactly. Can you spot what I am doing wrong?

**Comment by alexcrichton** [comment]
Thanks for the report! The issue here is that the WebAssembly module you're producing is importing functions from the host, namely WASI, but you're not supplying those imports to the instantiation of the module. That's the `[]` argument in the `Instance` constructor.

What I'd recommend you use instead is the `Linker` type which has a `define_wasi` function to define all WASI functions as-provided by Wasmtime. Additionally you'll then want to call `Store.set_wasi` with your WASI configuration which is where you're configure the argv which in your case will be `['./module.wasm', '2022-01-01']`

**Comment by alexcrichton** [comment]
Ah so unfortunately your Go code is mixed up. There's both a `main` function and a `parse` exported function. When you run the `wasmtime` CLI it runs the `main` function, not `parse` directly (although `main` calls `parse`). In the Python you're trying to call the `parse` function directly. The `parse` function, however, takes a `string` parameter which is more-or-less not possible to do like that. WebAssembly modules can only export functions that takes integers/floats as parameters so while you wrote a `string`-taking function the actual wasm (if you inspect the text format) is probably something like a function that takes two `i32` parameters. You're invoking the function, however with zero wasm parameters with `run(store)`. 

Invoking `parse` with a string is significantly more complicated than you might expect. I don't know how TinyGo lays out strings to do this. In other languages this is done by the embedder calling a `malloc` function, filling in the bytes manually, and then invoking `parse` with the return pointer from `malloc`. The `parse` function, however, then needs to understand memory management and everything.

Basically when you're working with core wasm you're working at a very low level. Features like the component model lift the layer of abstraction to something you're likely more used to, but there is not support just yet to easily use TinyGo with the component model.

**Comment by thisisaaronland** [comment]
Ah, got it. Thanks for the explanation.
